fix(api): more intuitive cursor updates in nvim_buf_set_text

faergeek opened the pull request:
Hello there! I'm trying to implement a bit more intuitive cursor position updates in `nvim_buf_set_text`. I looked through the issues and found one which mentions just one case https://github.com/neovim/neovim/issues/22526, which I as well find a bit counterintuitive, though it makes sense in INSERT mode as @bfredl clarified. I left a [comment](https://github.com/neovim/neovim/issues/22526#issuecomment-1693906375) there, suggesting that behavior could probably be different depending on the current mode/window.

I also described the problem I'm trying to solve within a plugin for code formatting. In short, moving cursors to proper positions after code formatting is a really nice feature, as well as adjusting extmark positions. Fortunately, `nvim_buf_set_text` already does all of that, though not in every case. And I think it would be great if `nvim_buf_set_text` would make of those kinds of adjustments. Based on the issue https://github.com/neovim/neovim/issues/10096 that seems to be one of the main reasons this api was introduced in the first place.

Right now I added a couple of test cases (there definitely should be more of them), in which `nvim_buf_set_text` could behave better IMO. And I only implemented a very basic change which roughly addresses first two of them (breaking the next two though :-)). So before I go further, I would like to know if these kinds of changes would be accepted at all, because they are kind of breaking. For example, right now lsp's `apply_text_edits` tests don't pass already with these changes. But I hope that may be in the end explicit cursor adjustments in `apply_text_edits` would not be needed at all, who knows... Otherwise, I'll just dig in and fix the lua code as well.

@bfredl could you please take a look at the test cases I added and tell me if I should continue fulfilling them and adding more as well?

bfredl commented:
Thanks for looking into this. I also noticed the current behavior was suboptimal, even though that was a separate issue from what currently regressed and will be fixed. I'll do a proper review soon.

bfredl commented on src/nvim/api/buffer.c:
```diff
@@ -722,8 +722,16 @@ void nvim_buf_set_text(uint64_t channel_id, Buffer buffer, Integer start_row, In
-      if (win->w_cursor.lnum == end_row && win->w_cursor.col > end_col) {
-        win->w_cursor.col -= col_extent - (colnr_T)last_item.size;
+      if (win->w_cursor.lnum == end_row) {
+        if (win == curwin && (State & MODE_INSERT)) {
```
I think this makes sense. in extmark terms, the insert mode cursor has different "gravity" than normal (and non-current win) cursor. 

however instead of code duplication I suggest write some thing like `int insert_adjust = (win == curwin && (State & MODE_INSERT)) ? 1 : 0` and then use that adjustment in a non-duplicated if statement.

faergeek commented:
@bfredl Thanks for comments :+1:

I suppose there's generally no objections regarding the idea, right?

Should I convert PR to draft until I implement it?

bfredl commented:
> I suppose there's generally no objections regarding the idea, right?
> Should I convert PR to draft until I implement it?

There is no general objection. Go head and implement the behavior you think reasonable, and we can discuss it from there.

faergeek commented:
@bfredl I updated the code. Could you please take a look again?

Now `col_extent` is only being used for fixing extmarks, logic for cursors does not depend on it's value anymore. Logic for extmarks should probably be changed too, though I'm not sure how exactly. When I tried to blindly apply the same logic, test cases related to extmarks started to fail. IIRC from reading the docs some time ago, it's intended that extmarks behave differently than what you'd expect from cursors. Plus I couldn't figure out what exactly `extmark_splice` is doing inside and what meaning `col_extent` really has there.

And as I expected, most tests for `apply_text_edits` pass even though I removed additional code that was fixing cursors there. One case is still failing though and I'm not sure what to do about it. It fails exactly because of "cursor gravity" changes in NORMAL mode. There's an edit being applied, which deletes characters precisely from 0 to cursor column, in which case cursor is shifted to 0 now. Before that change it would not move. I looked at git blame and found that this test was introduced in this PR https://github.com/neovim/neovim/pull/16038. However it seems like it's more concerned about cursor row falling out of bounds, not about column. @hrsh7th do you happen to remember if that was the only reason to add that test case?

BTW, do I understand correctly that internally cursor column value is expressed in bytes, not in characters?

faergeek commented:
I just rebased on master and changed column in `apply_text_edits` test for now. Not sure if this test should actually be deleted instead.

bfredl commented:
I agree with the changes with respect to the raw API calls, but I am not well versed with the LSP logic to make any conclusion on my own just yet. 

One note regarding nvim_buf_set_text and insert vs normal mode: it might be tempting to use set_text _intentionally_ to insert text at the cursor, but I think we should add a recommendation to use nvim_put at the doc string for this use case, which lets the plugin directly specify if the cursor should be placed before or after the insert cursor ("communicate intent precisely") .

faergeek commented:
> I agree with the changes with respect to the raw API calls, but I am not well versed with the LSP logic to make any conclusion on my own just yet.

Regarding LSP logic. If I flip the logic in these changes a bit (in short, mostly by replacing `-`s with `+`s) it would be much more clear that these changes in `nvim_buf_set_text` almost exactly mirror what was in `apply_text_edits`. And IMHO it would be much easier to understand what happens. Right now I left it like this mostly because it seems like similar changes should be applied to extmarks as well, if I figure out how.

From the code it seems like `col_extent` represents how the width of last line has changed. Is that what was intended? Or am I completely wrong? Sorry, I still couldn't figure out what happens in `extmark_splice`.

> One note regarding nvim_buf_set_text and insert vs normal mode: it might be tempting to use set_text _intentionally_ to insert text at the cursor, but I think we should add a recommendation to use nvim_put at the doc string for this use case, which lets the plugin directly specify if the cursor should be placed before or after the inserted text ("communicate intent precisely") .

You mean something similar to an existing note "Prefer |nvim_buf_set_lines()| if you are only adding or deleting entire lines."?

faergeek commented:
I tried adding similar test cases for extmarks and it seems like they are adjusted properly. So I guess nothing extra is needed there actually. Sorry for wasting time with this.

faergeek commented:
Rebased on master and added a comment recommending `nvim_put`. Now I'm also thinking about adjusting marks (not extmarks), since right now IIUC they are completely removed from lines affected by `nvim_buf_set_text`

hrsh7th commented:
> And as I expected, most tests for apply_text_edits pass even though I removed additional code that was fixing cursors there. One case is still failing though and I'm not sure what to do about it. It fails exactly because of "cursor gravity" changes in NORMAL mode. There's an edit being applied, which deletes characters precisely from 0 to cursor column, in which case cursor is shifted to 0 now. Before that change it would not move. I looked at git blame and found that this test was introduced in this PR https://github.com/neovim/neovim/pull/16038. However it seems like it's more concerned about cursor row falling out of bounds, not about column. @hrsh7th do you happen to remember if that was the only reason to add that test case?

Yes. My intention was just "rows going out of bounds".
I think you are correct in saying "After deleting the text before the cursor, the cursor should be moved to column 0".

faergeek commented:
I added a couple of tests (failing right now) for cases when cursor is inside replaced range. Seems like such behavior would make sense. And I think it would also make sense to adjust marks the same way. And then additional mark adjustments in `apply_text_edits` could be deleted as well.

@folke I see that this mark adjustment was done by you. Can you please take a look at these tests? Basically I think it would make sense if cursors/marks would be kept either at the same exact position they were before or at least inside replaced region if it got smaller.

faergeek commented:
Bumped into a case where cursor column can end up after last line column. This only happens when calling `nvim_win_get_cursor` right after `nvim_buf_set_text`, not caught by current tests. Plus I'm not sure what to do if `virtualedit` is on.

faergeek commented:
I added tests for cases where cursor column could go out of bounds and made them pass

faergeek commented:
@bfredl what do you think about adjusting marks in the same way? I see there's a `mark_col_adjust` function and at the first glance it seems like a separate function can be extracted from it, which would accept `buf` as argument. And it could be used instead of `mark_adjust_buf`. But at the same time I see that `diff_mark_adjust` is only called from `mark_adjust_buf` and I'm not yet sure why. Also recently there were changes related to topline adjustments and IIUC they were only made to `mark_adjust_buf`. Would it make sense to extract one more function from `mark_col_adjust` and go that way or is there a better way? Or could you give any advice on how to approach this?

Also, `nvim_buf_set_text` and `nvim_buf_set_lines` share a lot of similarities. Would it make sense to reimplement `nvim_buf_set_lines` in terms of `nvim_buf_set_text`, so that `nvim_buf_set_lines` would benefit from those adjustments to cursors and probably marks as well?

bfredl commented:
"Normal" named marks is a bit of tricky question as there is a lot of editing operations which don't adjust them per column. `mark_col_adjust` is applied to some operations, like split/join, but not most of them.

I'd say we consider that out of scope for this PR. If there is a simple fix to make it work with `nvim_buf_set_text` (as there is just four callsites for `mark_col_adjust`, adding a buf_T param in place is just fine), we can do that as a follow up PR, but I'll hold off larger refactors until we have figured out a sensible design. I e if we really want named marks to be adjusted like extmarks, we should do that consequentially by adjusting them in the extmark_adjust codepath. 

> Also, nvim_buf_set_text and nvim_buf_set_lines share a lot of similarities. Would it make sense to reimplement nvim_buf_set_lines in terms of nvim_buf_set_text, so that nvim_buf_set_lines would benefit from those adjustments to cursors and probably marks as well?

I don't think that's needed or desired. set_lines should be regarded as intentionally replacing entire lines, disregarding column state. It's better for _plugins_  to switch from set_lines to set_text intentionally when some columns of the lines should be preserved (encourage intent to be directly communicated, not inferred).

bfredl commented on src/nvim/api/buffer.c:
```diff
@@ -725,11 +727,47 @@ void nvim_buf_set_text(uint64_t channel_id, Buffer buffer, Integer start_row, In
-      if (win->w_cursor.lnum == end_row && win->w_cursor.col > end_col) {
-        win->w_cursor.col -= col_extent - (colnr_T)last_item.size;
+      if (win->w_cursor.lnum >= start_row && win->w_cursor.lnum <= end_row) {
+        colnr_T mode_col_adj = win == curwin && (State & MODE_INSERT) ? 0 : 1;
+        colnr_T cursor_gravity_col = win->w_cursor.col + mode_col_adj;
+
+        if (win->w_cursor.lnum == end_row && cursor_gravity_col > end_col) {
+          win->w_cursor.lnum += (linenr_T)extra;
+          win->w_cursor.col += (colnr_T)last_item.size - (colnr_T)end_col;
+
+          if (new_len == 1) {
+            win->w_cursor.col += (colnr_T)start_col;
+          }
+        } else {
+          linenr_T repl_end_row = (linenr_T)start_row + (linenr_T)replacement.size - 1;
+          bool cursor_was_after_end_row = win->w_cursor.lnum > repl_end_row;
+
+          if (cursor_was_after_end_row) {
+            win->w_cursor.lnum = repl_end_row;
+          }
+
+          colnr_T repl_start_col = new_len == 1 ? (colnr_T)start_col : 0;
+          int index = win->w_cursor.lnum - (int)start_row;
+          String repl_item = replacement.items[index].data.string;
+          colnr_T repl_end_col = repl_start_col + (colnr_T)repl_item.size;
+
+          if (cursor_was_after_end_row
+              || (win->w_cursor.lnum == repl_end_row
+                  && win->w_cursor.col > repl_end_col)) {
+            win->w_cursor.col = repl_end_col;
+
+            if (win->w_cursor.col - mode_col_adj >= repl_start_col) {
+              win->w_cursor.col -= mode_col_adj;
+            }
+          }
+        }
+
+        check_cursor_col_win(win);
+        changed_cline_bef_curs(win);
+        invalidate_botline(win);
```
I'd suggest to break out this into a new `fix_cursor_cols` function, even if it is just called in one place.

bfredl commented:
> Is there anything else I should do in this PR, besides figuring out how cursors should behave with virtualedit?

That's sounds alright.

faergeek commented:
Extracted `fix_cursor_cols` function and added `virtualedit` handling. Also added comments, not sure if it's too much)

justinmk commented on src/nvim/api/buffer.c:
```diff
@@ -504,6 +504,8 @@ void nvim_buf_set_lines(uint64_t channel_id, Buffer buffer, Integer start, Integ
 ///
+/// Prefer |nvim_put()| if you want to insert text at the cursor position.
+///
 /// @see |nvim_buf_set_lines()|
```
/// @see |nvim_put()|

faergeek replied:
Added that as well, thanks

justinmk commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -454,23 +454,6 @@ function M.apply_text_edits(text_edits, bufnr, offset_encoding)
-        col = -1,
-      }
-    end
-    local cursor = api.nvim_win_get_cursor(0)
```
very nice to see this deleted code!

zeertzjq commented on src/nvim/api/buffer.c:
```diff
@@ -1339,6 +1342,79 @@ static void fix_cursor(win_T *win, linenr_T lo, linenr_T hi, linenr_T extra)
   invalidate_botline(win);
 }
 
+// Fix cursor position after replacing text
+// between (start_row, start_col) and (end_row, end_col).
+//
+// win->w_cursor.lnum is assumed to be >= start_row and <= end_row.
```
```suggestion
/// Fix cursor position after replacing text
/// between (start_row, start_col) and (end_row, end_col).
///
/// win->w_cursor.lnum is assumed to be >= start_row and <= end_row.
```

faergeek replied:
I applied this change. Though I'm not sure what this is for. I thought `///` is used for proper docstrings for an api. Does it make a difference for an internal function as well?

bfredl replied:
yes, also internal functions can have docstrings (as documentation for core development)

faergeek committed:
fix(api): more intuitive cursor updates in nvim_buf_set_text

Fixes #22526

bfredl commented:
Great work, thanks.

And the effort put into tests is greatly appreciated. will make reasoning and refactoring this code much easier in the future.
